src/gadgets/sk_knowledge/sk_know_gadget.rs

```diff
@@ -1,22 +1,122 @@
+        Y: LinearCombination::from(Scalar::one()),
+        Z: LinearCombination::from(Scalar::one()),
+        T: LinearCombination::from(Scalar::zero()),
+    };
+    // Compute pk'
+    sk.reverse();
+    for var in sk {
+        // Check that var is either `0` or `1`
+        binary_constrain_gadget(cs, var);
+        Q = Q.double(cs);
+        // If bit == 1 -> Q = Q + basep
+        let basep_or_id = basep.conditionally_select(LinearCombination::from(var), cs);
+        Q = Q.add(cs, basep_or_id);
     }
-    // Constraint final result of Q against Pk.
+    // Constraint pk' == pk
+    pk.equals(cs, Q);
 }
-// Poin Addition P1 + P2 = P3
-// Point Doubling 2*P1 = P2
-// Conditionally selection -> 1 | 0 -> Do somethig
 
-// 1 0 0 1 1 1 0 1 ...... 0 1 1 0
+fn sk_knowledge_proof(
+    pc_gens: &PedersenGens,
+    bp_gens: &BulletproofGens,
+    basep: SonnyRistrettoPoint,
+    pk: SonnyRistrettoPoint,
+    sk: &[Scalar],
+) -> Result<(R1CSProof, Vec<CompressedRistretto>), R1CSError> {
+    // Generate transcript
+    let mut transcript = Transcript::new(b"Sk_knowledge");
+    // Generate prover
+    let mut prover = Prover::new(pc_gens, &mut transcript);
+    // Commit high-level variables
+    let (commitments, bits): (Vec<CompressedRistretto>, Vec<Variable>) = sk
+        .iter()
+        .map(|x| prover.commit(*x, Scalar::random(&mut rand::thread_rng())))
+        .unzip();
+
+    // Apply sk_knowledge_gadget
+    let basep_gadget = SonnyRistrettoPointGadget::from_point(basep, &mut prover);
+    let pk_gadget = SonnyRistrettoPointGadget::from_point(pk, &mut prover);
+    sk_knowledge_gadget(&mut prover, basep_gadget, pk_gadget, bits);
+
+    // Generate the proof
+    let proof = prover.prove(bp_gens)?;
+    Ok((proof, commitments))
+}
+
+fn sk_knowledge_verify(
+    pc_gens: &PedersenGens,
+    bp_gens: &BulletproofGens,
+    basep: SonnyRistrettoPoint,
+    pk: SonnyRistrettoPoint,
+    sk_bits_comms: Vec<CompressedRistretto>,
+    proof: &R1CSProof,
+) -> Result<(), R1CSError> {
+    // Generate transcript
+    let mut transcript = Transcript::new(b"Sk_knowledge");
+    // Generate verifier
+    let mut verifier = Verifier::new(&mut transcript);
+    // Commit high-level variables
+    let sk_bit_vars: Vec<Variable> = sk_bits_comms.iter().map(|x| verifier.commit(*x)).collect();
+    // Apply sk_knowledge_gadget
+    let basep_gadget = SonnyRistrettoPointGadget::from_point(basep, &mut verifier);
+    let pk_gadget = SonnyRistrettoPointGadget::from_point(pk, &mut verifier);
+    sk_knowledge_gadget(&mut verifier, basep_gadget, pk_gadget, sk_bit_vars);
+    // Verify the proof
+    verifier.verify(proof, pc_gens, bp_gens, &mut rand::thread_rng())
+}
+
+fn sk_knowledge_gadget_roundtrip_helper(
+    basep: SonnyRistrettoPoint,
+    pk: SonnyRistrettoPoint,
+    sk: SonnyScalar,
+) -> Result<(), R1CSError> {
+    let pc_gens = PedersenGens::default();
+    let bp_gens = BulletproofGens::new(8192, 1);
```

Comment:
8192 is strictly required? We are using 4096 for Phoenix

Reply:
Yes, we can't do it with less multiply gates.. Maybe #10 reduces the amount of gates, but atm this is needed..

Reply:
Theres aprox 7000 of them